src/main/java/randoop/generation/Detective.java

```diff
@@ -0,0 +1,332 @@
+   *     parts of Randoop
+   * @param secondaryObjPool the secondary object pool that stores sequences generated by Detective
+   * @param t the class type for which the input objects need to be constructed
+   * @return a SimpleList of method sequences that produce objects of the required type
+   */
+  public static SimpleList<Sequence> demandDrivenInputCreation(
+      ObjectPool mainObjPool, ObjectPool secondaryObjPool, Type t) {
+    // Extract all constructors/methods that constructs/returns the demanded type by
+    // searching through all methods of the main object pool.
+    Set<TypedOperation> producerMethodSet = extractProducerMethods(t);
+
+    // For each producer method, create a sequence that produces an object of the demanded type
+    // if possible, or produce a sequence that leads to the eventual creation of the demanded type.
+    for (TypedOperation producerMethod : producerMethodSet) {
+      Sequence newSequence = getInputAndGenSeq(mainObjPool, secondaryObjPool, producerMethod);
+      if (newSequence != null) {
+        // Execute the sequence and store the resultant object in the secondary object pool
+        // if the sequence is successful.
+        processSuccessfulSequence(secondaryObjPool, newSequence);
+      }
+    }
+
+    // Extract all method sequences that produce objects of the demanded type from the secondary
+    // object pool and return them.
+    return extractCandidateMethodSequences(secondaryObjPool, t);
+  }
+
+  /**
+   * Identifies a set of methods that construct objects of a specified type.
+   *
+   * <p>The method checks for all visible methods and constructors in the specified type that return
+   * the same type. It also recursively searches for inputs needed to execute a method that returns
+   * the type. The recursive search terminates if the current type is a primitive type or if it has
+   * already been processed.
+   *
+   * @param t the type for which the producer methods are to be extracted
+   * @return a set of TypedOperations that construct objects of the specified type t
+   */
+  public static Set<TypedOperation> extractProducerMethods(Type t) {
+    Set<Type> processed = new HashSet<>();
+    Queue<Type> workList = new ArrayDeque<>();
+
+    // The set of producer methods that construct objects of the specified type.
+    Set<TypedOperation> producerMethodSet = new LinkedHashSet<>();
+
+    // The set of types that are needed to construct objects of the specified type.
+    Set<Type> producerTypeSet = new HashSet<>();
+    workList.add(t);
+
+    // Recursively search for methods that construct objects of the specified type.
+    while (!workList.isEmpty()) {
+      Type currentType = workList.poll();
+
+      // Only consider the type if it is not a primitive type or if it hasn't already been
+      // processed.
+      if (!processed.contains(currentType) && !currentType.isNonreceiverType()) {
```

Comment:
The demand-driven approach does call methods that are not specified in `--classlist`, and it does contradict with that part of the manual. The paper's intention for this component was to create the missing input that requires external libraries for a method under test, which is something vanilla Randoop cannot do. When an input object of a type is missing for a method under test, the demand-driven approach creates it so the method can still be tested, regardless of whether the methods used to construct the inputs is from the set of classes supplied to Randoop - the priority for the component is to create the missing input so that the method can be actually tested, whether it is from an external library or from a part of the same program as the method under test that isn't under test. So when demand-driven testing is used, the statement "Randoop's tests only use classes you specify" is no longer necessarily true. If I am correct on what I had, I think it might be a good idea to talk about this specific use case in the manual.
Another thing the demand-driven approach sort of automatically handles is related to this section of the manual:
_"Suppose you want to test method A.m. You should direct Randoop to call not just A.m, but also methods that create A.m's inputs."_
This human effort of specifying methods to use does not feel right (from my limited understanding) as a default practice because it can become complex and tedious as the program scales. If demand-driven is integrated into Randoop, it may be more commonly used than manually specifying all methods that create the input for the method under test, as demand-driven is sufficient for getting the method tested, and manual method specification becomes a way to have more control over how the method's inputs are created. If I am correct on this, it should add to my arguments that this part of the code stays the same for now. I will think about how to mention this in the manual.